### DISCUSSION TITLE
Can I wrap .so library in kotlin native and call it from JVM using Knee?

### DISCUSSION THREAD

**Discussion by lemcoder** [deepmedia/Knee#6] [Q&A] [answered]
I am wondering if it would be possible to wrap a native shared library (for example a .so) using kotlin native and CInterop and then call that code from JVM using Knee? I mean something like that:

some native lib (.so) -> cinterop -> kotlin native -> knee -> kotlin/jvm

This would greatly improve the quality of working with native libraries from multiplatform code 🙌

**Comment by natario1** [comment, marked as answer]
If you manually write wrapper functions/classes and annotate them properly, yes.

> This would greatly improve the quality of working with native libraries from multiplatform code 🙌

If you are writing a new application, you can strive to have as much logic as possible in the Kotlin/Native layer, and consume your `*.so` libraries there (no Knee needed). Then, use Knee to expose data to your UI. Basically, I would try to have `androidMain` as small as possible, only the UI and a few helpers to access platform APIs that are missing in the NDK. Put most stuff in `androidNativeMain` (or `nativeMain`, shared with iOS).

**Reply by adamglin0** [reply]
After my reading of the document, is it only androidNative that can do this?

_Since I am a beginner with JNI, although I would like to assist, I am unable to._

**Reply by natario1** [reply]
You can track #7 for full native targets support
